File: 33_container_with_most_water/maxarea.py
class Solution:
    # @return an integer
    def maxArea(self, height):
        """
        size = len(height)
        max_area = [0] * size
        #return sum(height)
        for i in xrange(size):
            for j in xrange(i):
                area = min(height[i],height[j])*(i-j)
                if area > max_area[i]:
                    max_area[i] = area
        return max(max_area)
        """
        left = 0
        right = len(height)-1
        result = 0
        while left < right:
            result = max(result, (right-left)*min(height[left], height[right]))
            if height[left] < height[right]:
                left +=1
            else:
                right -=1
        return result

if __name__ == "__main__":
    solution = Solution()
    height = [2,4,8,4,5]
    print solution.maxArea(height)

File: 33_container_with_most_water/maxarea.cpp
#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    int maxArea(vector<int> &height) {
        int size = height.size();
        int left = 0;
        int right = size - 1;
        int result = 0;
        while (left < right) {
            int left_val = height[left];
            int right_val = height[right];
            int cur_area = (right - left) * (left_val < right_val ? left_val : right_val);
            if (cur_area > result)
                result = cur_area;
            if (left_val < right_val)
                left += 1;
            else
                right -= 1;
        }
        return result;
    }
};

int main() {
    Solution solution;
    vector<int> height;
    int h_array[] = {2,4,8,4,5};
    for (int i = 0; i < sizeof(h_array)/sizeof(int); i++)
        height.push_back(h_array[i]);
    cout << solution.maxArea(height) << endl;
    return 0;
}
